File: cmdutils.py
PIPE = 124
FILE = 62
COLN = 59
EVAL = 69

FILEOUT = 4962
FILEERR = 5062

def tokenize(string):
    chars = list(string.strip())
    i = 0
    tokens = []
    temp = []
    
    while i < len(chars):
        c = chars[i]
        
        if c == '\\':
            temp.append(chars[i + 1])
            i += 1
        
        elif c in ('>', '|', ';'):
            if temp:
                tokens.append(''.join(temp))
                temp = []
            tokens.append(ord(c))
            
        elif c == ' ':
            if temp:
                tokens.append(''.join(temp))
                temp = []
                
        elif c in ("'", '"'):
            i += 1
            try:
                if chars[i] == c:
                    temp.append(c)
                    temp.append(c)
                else:
                    while chars[i] != c:
                        temp.append(chars[i])
                        i += 1
            except IndexError as err:
                raise err
            
        else:
           temp.append(c)
        
        i += 1
    
    if j := ''.join(temp):
        tokens.append(''.join(temp))

    if tokens[-1] != COLN:
        tokens.append(COLN)

    return tokens

def tree(tokens):
    temp = []
    tree = []
    i = 0
    
    if tokens[-1] != COLN:
        tokens.append(COLN)
    
    while i < len(tokens):
        if tokens[i] in (PIPE, COLN, FILE, FILEOUT, FILEERR):
            if temp:
                tree.append(temp)
                temp = []
            tree.append(tokens[i])
        else:
            temp.append(tokens[i])
        
        i += 1
    
    return tree

def parse(tokens):
    i = 0
    colncount = 0
    pipecount = 0
    filecount = 0
    instrs = []
    
    while i < len(tokens):
        if tokens[i] == COLN:
            assert type(tokens[i - 1]) == list
            
            instrs.append([COLN])
            i += 1
            
            colncount += 1
            
        elif tokens[i] == PIPE:
            assert type(tokens[i + 1]) == list
            
            instrs.append([PIPE, tokens[i + 1]])
            i += 2
            
            pipecount += 1
            
        elif tokens[i] in (FILE, FILEOUT, FILEERR):
            assert type(tokens[i + 1]) == list
            
            instrs.append([tokens[i], tokens[i + 1][0]])
            i += 2
            
            filecount += 1 
            
        else:
            instrs.append([EVAL, tokens[i]])
            i += 1
        
    return instrs, (colncount, pipecount, filecount)
            
    
    
    
    
    

File: speedups/src/cmdutils.rs
pub enum Op {
    Pipe,
    File,
    FileOut,
    FileErr,
    Coln,
}

pub enum Instrs {
    Pipe,
    File,
    FileOut,
    FileErr,
    Coln,
}

#[derive(Debug)]
pub enum Token {
    PartialOp(String),
    Str(String),
}

#[derive(Debug)]
pub enum TokenizerError {
    StringScanEOF,
    EscapeSeqEOF,
    PatScanEOF,
    PatZeroLength,
}

pub fn tokenize(chars: Vec<char>) -> Result<Vec<Token>, TokenizerError> {
    let mut i = 0;
    let mut tokens: Vec<Token> = Vec::new();
    let mut temp: Vec<char> = Vec::new();

    while i < chars.len() {
        match chars[i] {
            '\\' => {
                i += 1;
                if !(i < chars.len()) {
                    return Err(TokenizerError::EscapeSeqEOF);
                }
                
                if chars[i] == 'n' {
                    temp.push('\n');
                } else {
                    temp.push(chars[i]);
                }
            }

            '>' | '|' | ';' => {
                let mut s = temp.iter().collect::<String>();

                // this is to support 1> and 2>
                // all other pairs will be considered as seperate tokens
                if (chars[i] == '>'/*|| chars[i] == '|'*/) && (s == "1" || s == "2") {
                    tokens.push(Token::PartialOp(format!("{}{}", s, chars[i]).to_string()));
                } else {
                    if s.len() != 0 {
                        tokens.push(Token::Str(s));
                    }
                    tokens.push(Token::PartialOp(chars[i].to_string()));
                }

                temp.clear();
            }

            ' ' | '\n' => {
                if temp.len() != 0 {
                    tokens.push(Token::Str(temp.iter().collect::<String>()));
                    temp.clear();
                }
            }

            '"' | '\'' => {
                let c = chars[i].clone();
                i += 1;
                if !(i < chars.len()) {
                    return Err(TokenizerError::StringScanEOF);
                }

                if chars[i] == c {
                    // basically if the string is '' or "" then
                    // keep the quotes too
                    temp.push(c.clone());
                    temp.push(c.clone());
                } else {
                    while chars[i] != c {
                        temp.push(chars[i].clone());
                        i += 1;

                        if !(i < chars.len()) {
                            return Err(TokenizerError::StringScanEOF);
                        }
                    }
                }
            }

            '*' => {
                // with this you can pass text with both quotes and newlines
                /* example:
                echo *EOF
                    print('hello')
                EOF

                also works with codeblocks:

                echo *```py
                    print('hello')
                ```
                */

                let mut pat = String::new();
                loop {
                    i += 1;
                    if !(i < chars.len()) {
                        return Err(TokenizerError::PatScanEOF);
                    }
                    match chars[i] {
                        'a'..='z' | 'A'..='Z' | '0'..='9' | '`' => {
                            pat.push(chars[i].clone());
                        }
                        _ => {
                            break;
                        }
                    }
                }

                if pat.len() == 0 {
                    return Err(TokenizerError::PatZeroLength);
                }

                pat = if pat.starts_with("```") {
                    "```".to_string()
                } else {
                    pat
                };

                let mut cont = String::new();
                while !cont.ends_with(&pat) {
                    match chars.get(i) {
                        Some(some) => cont.push(some.clone()),
                        None => return Err(TokenizerError::PatScanEOF),
                    }

                    i += 1;
                }
                i -= 1;

                let cont = cont.chars().collect::<Vec<char>>();
                temp.extend(&cont[..cont.len() - pat.len()]);
            }

            _ => {
                temp.push(chars[i].clone());
            }
        }
        i += 1;
    }

    if temp.len() != 0 {
        tokens.push(Token::Str(temp.iter().collect::<String>()));
    }

    return Ok(tokens);
}

/*fn main() {
    let mut chars = String::from(
r#"a > *EOF lol EOF"#).chars().collect::<Vec<char>>();

    dbg!(tokenize(chars));

}
*/
